Errors linking libhpx.so due to missing references (master branch, commit 6679a8882)

I was unable to build the above-noted version of HPX both on gcc (v. 7.2.0) and icc (18.0.1).  My development workstation (running Debian testing):
Linux dev 4.13.0-1-amd64 #1 SMP Debian 4.13.13-1 (2017-11-16) x86_64 GNU/Linux

With both compilers, the build failed when linking libhpx.so (I did not try with clang this time).  The error messages (similar for both) as follows:

```
[ 4%] Linking CXX shared library ../lib/libhpx.so
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::load(unsigned __int128 const volatile&, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::load(unsigned __int128 const volatile&, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::load(unsigned __int128 const volatile&, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::load(unsigned __int128 const volatile&, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::load(unsigned __int128 const volatile&, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::load(unsigned __int128 const volatile&, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o:/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: more undefined references to `__atomic_load_16' follow
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::load(unsigned __int128 const volatile&, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:100: undefined reference to `__atomic_load_16'
CMakeFiles/hpx.dir/runtime/threads/detail/scheduled_thread_pool.cpp.o: In function `boost::atomics::detail::gcc_atomic_operations<unsigned __int128>::compare_exchange_strong(unsigned __int128 volatile&, unsigned __int128&, unsigned __int128, boost::memory_order, boost::memory_order)':
/usr/src/boost/1.64/gcc/boost/atomic/detail/ops_gcc_atomic.hpp:126: undefined reference to `__atomic_compare_exchange_16'
collect2: error: ld returned 1 exit status
src/CMakeFiles/hpx.dir/build.make:4559: recipe for target 'lib/libhpx.so.1.1.0' failed
make[2]: *** [lib/libhpx.so.1.1.0] Error 1
make[2]: Target 'src/CMakeFiles/hpx.dir/build' not remade because of errors.
CMakeFiles/Makefile2:911: recipe for target 'src/CMakeFiles/hpx.dir/all' failed
make[1]: *** [src/CMakeFiles/hpx.dir/all] Error 2
make[1]: Target 'all' not remade because of errors.
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
make: Target 'default_target' not remade because of errors.
make -kj8  16.46s user 2.31s system 100% cpu 18.597 total
```
I've had this problem in the past and I cannot recall how I solved it at the time.

After some searching and trial-and-error, I found that adding "-latomic" to the build flags fixes the problem.

--- a/cmake/HPX_AddConfigTest.cmake
+++ b/cmake/HPX_AddConfigTest.cmake
@@ -9,6 +9,8 @@
 
 set(HPX_ADDCONFIGTEST_LOADED TRUE)
 
+include(CheckLibraryExists)
+
 macro(add_hpx_config_test variable)
   set(options FILE EXECUTE)
   set(one_value_args SOURCE ROOT CMAKECXXFEATURE)
@@ -387,15 +389,20 @@ macro(hpx_check_for_cxx11_std_atomic)
   # Sometimes linking against libatomic is required for atomic ops, if
   # the platform doesn't support lock-free atomics.
 
+  # remove REQUIRED so that we don't generate a failure too early
+  # need to copy ARGN to list in order to remove REQUIRED
+  set (extra_macro_args ${ARGN})
+  list(REMOVE_ITEM extra_macro_args REQUIRED)
+
   # First check if atomics work without the library.
   add_hpx_config_test(HPX_WITH_CXX11_ATOMIC
-    SOURCE cmake/tests/cxx11_std_atomic.cpp
-    FILE ${ARGN})
+    SOURCE cmake/tests/cxx11_std_atomic.cpp FILE extra_macro_args)
 
   # If not, check if the library exists, and atomics work with it.
   if(NOT HPX_WITH_CXX11_ATOMIC)
     check_library_exists(atomic __atomic_fetch_add_4 "" HPX_HAVE_LIBATOMIC)
     if(HPX_HAVE_LIBATOMIC)
+      unset(HPX_WITH_CXX11_ATOMIC CACHE)
       add_hpx_config_test(HPX_WITH_CXX11_ATOMIC
         SOURCE cmake/tests/cxx11_std_atomic.cpp
         LIBRARIES "atomic"

--- a/cmake/tests/cxx11_std_atomic.cpp
+++ b/cmake/tests/cxx11_std_atomic.cpp
@@ -6,9 +6,10 @@
 ////////////////////////////////////////////////////////////////////////////////
 
 #include <atomic>
+#include <cstdint>
 
 struct P {
-    long x; long y;
+    uint64_t x; uint64_t y;
 };
 
 int main()

--- a/cmake/tests/cxx11_std_atomic.cpp
+++ b/cmake/tests/cxx11_std_atomic.cpp
@@ -7,6 +7,10 @@
 
 #include <atomic>
 
+struct P {
+    long x; long y;
+};
+
 int main()
 {
     std::atomic_flag af = ATOMIC_FLAG_INIT;
@@ -16,6 +20,9 @@ int main()
     std::atomic<int> ai;
     ai.store(0);
     int i = ai.load();
+
+    std::atomic<P> p0;
+    P p1 = atomic_load(&p0);
 
     std::memory_order mo;
     mo = std::memory_order_relaxed;

--- a/cmake/tests/cxx11_std_atomic.cpp
+++ b/cmake/tests/cxx11_std_atomic.cpp
@@ -6,10 +6,9 @@
 ////////////////////////////////////////////////////////////////////////////////
 
 #include <atomic>
-#include <cstdint>
 
 struct P {
-    uint64_t x; uint64_t y;
+    long long x; long long y;
 };
 
 int main()

